### PR TITLE
Add support for blocking zpopmax, zpopmin commands

### PR DESCRIPTION
**Pull request by dengliming**
Resolves [https://github.com/redis/jedis/issues/2177](https://github.com/redis/jedis/issues/2177)

### CODE REVIEW

File: src/test/java/redis/clients/jedis/tests/commands/SortedSetCommandsTest.java [hunk 1]
```diff
@@ -1412,4 +1413,42 @@ public void infinity() {
...
+  public void bzpopmax() {
+    jedis.zadd("foo", 1d, "a", ZAddParams.zAddParams().nx());
+    jedis.zadd("foo", 10d, "b", ZAddParams.zAddParams().nx());
+    jedis.zadd("bar", 0.1d, "c", ZAddParams.zAddParams().nx());
+    ScoreValueTuple actual = jedis.bzpopmax(0, "foo", "bar");
+    assertEquals(new ScoreValueTuple("foo", "b", 10d), actual);
+
+    jedis.del("foo");
+    jedis.del("bar");
+
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/tests/commands/SortedSetCommandsTest.java:1427]: Not necessary.
```suggestion
    jedis.del("foo");
    jedis.del("bar");

```

File: src/test/java/redis/clients/jedis/tests/commands/SortedSetCommandsTest.java [hunk 2]
```diff
@@ -1412,4 +1413,42 @@ public void infinity() {
...
+  public void bzpopmin() {
+    jedis.zadd("foo", 1d, "a", ZAddParams.zAddParams().nx());
+    jedis.zadd("foo", 10d, "b", ZAddParams.zAddParams().nx());
+    jedis.zadd("bar", 0.1d, "c", ZAddParams.zAddParams().nx());
+    ScoreValueTuple actual = jedis.bzpopmin(0, "bar", "foo");
+    assertEquals(new ScoreValueTuple("bar", "c", 0.1d), actual);
+
+    jedis.del("foo");
+    jedis.del("bar");
+
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/tests/commands/SortedSetCommandsTest.java:1446]: Not necessary.
```suggestion
```

File: src/main/java/redis/clients/jedis/ScoreValueTuple.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    int result = super.hashCode();
+    result = 31 * result + (key != null ? Arrays.hashCode(key) : 0);
+    return result;
+  }
+
+  @Override
+  public String toString() {
+    return "ScoreValueTuple{" +
+            "key=" + SafeEncoder.encode(key) +
+            "} " + super.toString();
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/ScoreValueTuple.java:48]: This doesn't look right.

File: src/main/java/redis/clients/jedis/ScoreValueTuple.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    if (!super.equals(o)) return false;
+
+    ScoreValueTuple that = (ScoreValueTuple) o;
+    return Arrays.equals(key, that.key);
+  }
+
+  @Override
+  public int hashCode() {
+    int result = super.hashCode();
+    result = 31 * result + (key != null ? Arrays.hashCode(key) : 0);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/ScoreValueTuple.java:40]: IMO, it should be 31 * hash(key) + hash(tuple).

File: src/main/java/redis/clients/jedis/ScoreValueTuple.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
+package redis.clients.jedis;
+
+import redis.clients.jedis.util.SafeEncoder;
+
+import java.util.Arrays;
+
+public class ScoreValueTuple extends Tuple {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/ScoreValueTuple.java:7]: The name `ScoreValueTuple` doesn't feel right. The value score itself is actually contained in Tuple.

The best I can come up with right now is `KeyedTuple`. But fee free to use your imagination.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/ScoreValueTuple.java (hunk 5 above)
Comment (dengliming) [reply src/main/java/redis/clients/jedis/ScoreValueTuple.java]: Agree with `KeyedTuple`. Thanks.

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 6]
```diff
@@ -326,6 +326,24 @@ public String toString() {
 
   };
 
+  public static final Builder<KeyedTuple> SCORE_VALUE_TUPLE = new Builder<KeyedTuple>() {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BuilderFactory.java:329]:
```suggestion
  public static final Builder<KeyedTuple> KEYED_TUPLE = new Builder<KeyedTuple>() {
```

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 7]
```diff
@@ -326,6 +326,24 @@ public String toString() {
...
+      List<byte[]> l = (List<byte[]>) data; // never null
+      if (l.isEmpty()) {
+        return null;
+      }
+      return new KeyedTuple(l.get(0), l.get(1), DOUBLE.build(l.get(2)));
+    }
+
+    @Override
+    public String toString() {
+      return "ScoreValueTuple";
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BuilderFactory.java:342]:
```suggestion
      return "KeyedTuple";
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Review by sazzad16 — APPROVED** [review]

**Pull request merged**